### ISSUE TITLE
Fasttext Keras model import:  Unsupported keras layer type Embedding and GlobalAveragePooling1D

### ISSUE THREAD

**Issue by lefromage** [deeplearning4j/deeplearning4j#2686]
Hi, I am trying to import a Keras model from https://github.com/fchollet/keras/blob/master/examples/imdb_fasttext.py
I am getting an error: Unsupported keras layer type Embedding.

do you have any suggestion on how to circumvent.
do you have any other way to support fasttext in DL4J ?

Thanks

**Comment by lefromage** [comment]
the model json is fairly small:

{
  "class_name": "Sequential",
  "config": [
    {
      "class_name": "Embedding",
      "config": {
        "name": "embedding_1",
        "trainable": true,
        "batch_input_shape": [
          null,
          400
        ],
        "input_dtype": "int32",
        "input_dim": 20000,
        "output_dim": 50,
        "init": "uniform",
        "input_length": 400,
        "mask_zero": false,
        "activity_regularizer": null,
        "W_regularizer": null,
        "W_constraint": null,
        "dropout": 0
      }
    },
    {
      "class_name": "GlobalAveragePooling1D",
      "config": {
        "name": "globalaveragepooling1d_1",
        "trainable": true
      }
    },
    {
      "class_name": "Dense",
      "config": {
        "name": "dense_1",
        "trainable": true,
        "output_dim": 1,
        "init": "glorot_uniform",
        "activation": "sigmoid",
        "W_regularizer": null,
        "b_regularizer": null,
        "activity_regularizer": null,
        "W_constraint": null,
        "b_constraint": null,
        "bias": true,
        "input_dim": 50
      }
    }
  ],
  "keras_version": "1.2.0"
}

**Comment by agibsonccc** [comment]
Send us a pull request or wait for dave/susan to work on it? Dl4j already
supports an embedding layer so it shouldnt be too bad to add.

On Jan 16, 2017 08:03, "lefromage" <notifications@github.com> wrote:

Hi, I am trying to import a Keras model from https://github.com/fchollet/
keras/blob/master/examples/imdb_fasttext.py
I am getting an error: Unsupported keras layer type Embedding.

do you have any suggestion on how to circumvent.
do you have any other way to support fasttext in DL4J ?

Thanks

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/deeplearning4j/deeplearning4j/issues/2686>, or mute the
thread

**Comment by turambar** [comment]
@Lefromage @agibsoncc Yeah that should be an easy fix. I'll add it to the queue of mode import todos -- PR would be gratefully accepted!

That said, do we an analogue to Keras' global average pooling layer?

**Comment by AlexDBlack** [comment]
@turambar global pooling was added here: https://github.com/deeplearning4j/deeplearning4j/pull/2699

**Comment by agibsonccc** [comment]
This was fixed I believe? @turambar

**Comment by turambar** [comment]
@agibsonccc Yes, here: https://github.com/deeplearning4j/deeplearning4j/pull/2849. I haven't merged yet -- I have one more test to do (a Keras fasttext model to import). Once I finish that (tonight) I will merge and close this.

**Comment by turambar** [comment]
Fasttext (embedding + global pooling) is supported in DL4J and in Keras model import. Included in master and in upcoming release.

**Comment by vychen** [comment]
Importing into a MultiLayerNetwork and under version 0.7.2, I still see the error that Embedding Layer is not supported. Is there a different setting where embedding is supported? Thanks!


Exception in thread "main" org.deeplearning4j.nn.modelimport.keras.InvalidKerasConfigurationException: Unsupported keras layer type Embedding. For more information, see http://deeplearning4j.org/model-import-keras.
        at org.deeplearning4j.nn.modelimport.keras.KerasLayer.buildLayerFromConfig(KerasLayer.java:459)
        at org.deeplearning4j.nn.modelimport.keras.KerasLayer.<init>(KerasLayer.java:195)
        at org.deeplearning4j.nn.modelimport.keras.KerasModel.helperPrepareLayers(KerasModel.java:186)
        at org.deeplearning4j.nn.modelimport.keras.KerasSequentialModel.<init>(KerasSequentialModel.java:84)
        at org.deeplearning4j.nn.modelimport.keras.KerasSequentialModel.<init>(KerasSequentialModel.java:51)
        at org.deeplearning4j.nn.modelimport.keras.KerasModel$ModelBuilder.buildSequential(KerasModel.java:401)
        at org.deeplearning4j.nn.modelimport.keras.KerasModelImport.importKerasSequentialModelAndWeights(KerasModelImport.java:168)
        at com.ptrx.etl.features.KerasModel$.score(tmp.scala:18)
        at com.ptrx.etl.features.Main$.main(Main.scala:30)
        at com.ptrx.etl.features.Main.main(Main.scala)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:738)
        at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:187)
        at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:212)
        at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:126)
        at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)

**Comment by turambar** [comment]
@vychen Please read the entire comment thread, including my previous message:

> Included in master and in upcoming release (i.e., NOT 0.7.2).

So to get fasttext support, you'll need to either build from master or wait for the next release. And before you ask: no, we cannot tell you when that release is coming. ;) As a policy, we do not comment publicly on or make guarantees about release schedules for the open source distribution of DL4J.
